src/main/java/org/jboss/staxmapper/NaryVersion.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2023, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.staxmapper;
+
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * A version based on N terms.
+ * @author Paul Ferraro
+ */
+public class NaryVersion<T extends NaryVersion<T>> implements Comparable<T> {
+
```

Comment:
I've removed the self-type - which was meant to be defined by subtypes, e.g. BinaryVersion vs TernaryVersion; which are now scrapped.

I've added a method to override the comparator of a Versioned object for use by the since(...) method.  This enables the use of smallrye-common-version via something like:

```java
enum Version implements Versioned<String, MyNamespace> {
    V1_0("1.0"),
    V1_1("1.1"),
    ;
    private final String version;

    Version(String version) {
        this.version = version;
    }

    @Override
    public String getVersion() {
        return this.version;
    }

    @Override
    public Comparator<String> getComparator() {
        return VersionSchema.BASIC;
    }
}
```

While this works, I've kept IntVersion, which peforms faster comparisons.